generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt

```diff
@@ -0,0 +1,147 @@
+     * [progress] value.
+     */
+    private fun createForegroundInfo(progress: Int): ForegroundInfo {
+        // create a ForegroundInfo using the notificationId and a new progress notification
+        return ForegroundInfo(
+            notificationId,
+            workerNotification.createProgressNotification(progress)
+        )
+    }
+
+    override suspend fun doWork(): Result {
+        // get the job parameter which is the json file path
+        val offlineJobJsonPath = inputData.getString(jobParameter) ?: return Result.failure()
+        // load the json file
+        val offlineJobJsonFile = File(offlineJobJsonPath)
+        // if the file doesn't exist return failure
+        if (!offlineJobJsonFile.exists()) {
+            return Result.failure()
+        }
+        // create the GenerateOfflineMapJob from the json file
+        val generateOfflineMapJob =
+            GenerateOfflineMapJob.fromJson(offlineJobJsonFile.readText())
+            // return failure if the created job is null
+                ?: return Result.failure()
+        // a coroutine job reference for the generateOfflineMapJob's progress collector
+        var progressCollectorJob: Job? = null
+
+        return try {
+            // set this worker to run as a long-running foreground service
+            // this will throw an exception, if the worker is launched when the app
+            // is not in foreground
+            setForeground(createForegroundInfo(0))
+            // check and delete if the offline map package file already exists
+            // this check is needed, if the download has failed midway and is restarted later
+            // by WorkManager
+            File(generateOfflineMapJob.downloadDirectoryPath).deleteRecursively()
+
+            // start the generateOfflineMapJob
+            // this job internally runs on a Dispatchers.IO context, hence this CoroutineWorker
+            // can be run on the default Dispatchers.Default context
+            generateOfflineMapJob.start()
+
+            // launch a progress collector in a new CoroutineScope
+            progressCollectorJob = CoroutineScope(Dispatchers.Default).launch {
+                // collect on progress until the job has completed in a success/failure
+                generateOfflineMapJob.progress.takeWhile {
+                    generateOfflineMapJob.status.value != JobStatus.Failed
+                        && generateOfflineMapJob.status.value != JobStatus.Succeeded
+                }.collect { progress ->
+                    // update the worker progress
+                    setProgress(workDataOf("Progress" to progress))
+                    // update the ongoing progress notification
+                    setForeground(createForegroundInfo(progress))
+                }
+            }
+
+            // suspends until the generateOfflineMapJob has completed
+            val result = generateOfflineMapJob.result()
```

Comment:
using one of the coroutine scope functions to create a new scope is better than creating a new CoroutineScope instance directly here, since the use of a coroutine scope function ensures that the coroutine context is inherited from the parent, thus preserving structured concurrency. It also means you won't have to manage the `progressCollectorJob` with regards to cancellation.  I suggest the following:
```suggestion
            val result = coroutineScope {
                // launch progress collection in a new coroutine, for this we need a new coroutine scope, created with the coroutineScope() function
                launch {
                     // collect progress until the job has completed with success/failure
                     generateOfflineMapJob.progress.takeWhile {
                     generateOfflineMapJob.status.value != JobStatus.Failed
                        && generateOfflineMapJob.status.value != JobStatus.Succeeded
                     }.collect { progress ->
                      // update the worker progress
                      setProgress(workDataOf("Progress" to progress))
                      // update the ongoing progress notification
                      setForeground(createForegroundInfo(progress))
                  }
                }
                
                // suspends until the generateOfflineMapJob has completed
                generateOfflineMapJob.result()
            }
```